service/src/event/service.rs

```diff
@@ -101,53 +111,34 @@ impl CeramicEventService {
+    pub(crate) async fn validate_discovered_event(
+        event_id: ceramic_core::EventId,
         carfile: &[u8],
-    ) -> Result<(EventInsertableBody, Option<DeliverableMetadata>)> {
-        let insertable = EventInsertableBody::try_from_carfile(event_cid, carfile).await?;
-        let ev_block = insertable.block_for_cid(&insertable.cid)?;
-
-        trace!(count=%insertable.blocks.len(), cid=%event_cid, "parsing event blocks");
-        let event_ipld: unvalidated::RawEvent<Ipld> =
-            serde_ipld_dagcbor::from_slice(&ev_block.bytes).map_err(|e| {
-                Error::new_invalid_arg(
-                    anyhow::anyhow!(e).context("event block is not valid event format"),
-                )
-            })?;
-
-        let maybe_init_prev = match event_ipld {
-            unvalidated::RawEvent::Time(t) => Some((t.id(), t.prev())),
-            unvalidated::RawEvent::Signed(signed) => {
-                let link = signed.link().ok_or_else(|| {
-                    Error::new_invalid_arg(anyhow::anyhow!("event should have a link"))
-                })?;
-                let link = insertable.block_for_cid(&link).map_err(|e| {
-                    Error::new_invalid_arg(
-                        anyhow::anyhow!(e).context("prev CID missing from carfile"),
-                    )
-                })?;
-                let payload: unvalidated::Payload<Ipld> =
-                    serde_ipld_dagcbor::from_slice(&link.bytes).map_err(|e| {
-                        Error::new_invalid_arg(
-                            anyhow::anyhow!(e).context("Failed to follow event link"),
-                        )
-                    })?;
-
-                match payload {
-                    unvalidated::Payload::Data(d) => Some((*d.id(), *d.prev())),
-                    unvalidated::Payload::Init(_init) => None,
-                }
-            }
-            unvalidated::RawEvent::Unsigned(_init) => None,
-        };
-        let meta = maybe_init_prev.map(|(cid, prev)| DeliverableMetadata {
-            init_cid: cid,
-            prev,
-        });
-        Ok((insertable, meta))
+    ) -> Result<(EventInsertable, EventMetadata)> {
+        let event_cid = event_id.cid().ok_or_else(|| {
+            Error::new_app(anyhow::anyhow!("EventId missing CID. EventID={}", event_id))
+        })?;
+
+        let (cid, parsed_event) = unvalidated::Event::<Ipld>::decode_car(carfile, false)
+            .await
+            .map_err(Error::new_app)?;
+
+        if event_cid != cid {
+            return Err(Error::new_app(anyhow::anyhow!(
+                "EventId CID ({}) does not match the body CID ({})",
+                event_cid,
+                cid
+            )));
+        }
+
+        let metadata = EventMetadata::from(parsed_event);
+        let mut body = EventInsertableBody::try_from_carfile(cid, carfile).await?;
+        body.set_deliverable(matches!(metadata, EventMetadata::Init { .. }));
```

Comment:
Why is this logic here instead of part of the event ordering bits? While true any init event is deliverable it seems like this logic should live with the other logic that knows that an init event doesn't have a prev.

Reply:
I moved it to the Ordering task. Originally I thought it might be less likely to ever be missed if it just happened right away but I think it makes sense in the new location.